packages/hash/api/src/model/knowledge/entity.model.ts

```diff
@@ -112,6 +118,139 @@ export default class {
     });
   }
 
+  /**
+   * Create an entity along with any new/existing entities specified through links.
+   *
+   * @param params.createdById the account id that is creating the entity
+   * @param params.entityDefinition the definition of how to get or create the entity optionally with linked entities
+   */
+  static async createEntityWithLinks(
+    graphApi: GraphApi,
+    params: {
+      createdById: string;
+      entityDefinition: KnowledgeEntityDefinition;
+    },
+  ): Promise<EntityModel> {
+    const { createdById, entityDefinition: entityDefinitions } = params;
+
+    const entitiesInTree = linkedTreeFlatten<
+      KnowledgeEntityDefinition,
+      KnowledgeLinkedEntityDefinition,
+      "linkedEntities",
+      "entity"
+    >(entityDefinitions, "linkedEntities", "entity");
+
+    const entities = await Promise.all(
+      entitiesInTree.map(async (entityDefinition) => ({
+        link: entityDefinition.meta
+          ? {
+              parentIndex: entityDefinition.parentIndex,
+              meta: entityDefinition.meta,
+            }
+          : undefined,
+        entity: await EntityModel.getOrCreate(graphApi, {
+          createdById,
+          entityDefinition,
+        }),
+      })),
+    );
+
+    await Promise.all(
+      entities.map(async ({ link, entity }) => {
+        if (link) {
+          const parentEntity = entities[link.parentIndex];
+          if (!parentEntity) {
+            throw new ApolloError("Could not find parent entity");
+          }
+          const linkTypeModel = await LinkTypeModel.get(graphApi, {
+            versionedUri: link.meta.linkTypeVersionedUri,
+          });
+
+          // links are created as an outgoing link from the parent entity to the children.
+          await parentEntity.entity.createOutgoingLink(graphApi, {
+            linkTypeModel,
+            targetEntityModel: entity,
+            index: link.meta.index ?? undefined,
+          });
+        }
+      }),
+    );
+
+    if (entities.length > 0) {
+      // First element will be the root entity.
+      return entities[0]!.entity;
```

Comment:
Nit: we can avoid casting here
```suggestion
    if (entities[0]) {
      // First element will be the root entity.
      return entities[0].entity;
```